File: shop_editor.py
import os
import re
import tkinter as tk
from tkinter import ttk, filedialog, messagebox

try:  # optional Pillow dependency
    from PIL import Image, ImageTk
except Exception:  # Pillow missing or failed to load
    Image = None
    ImageTk = None

CELL_SIZE = 32
GRID_W = 8
GRID_H = 15


def parse_item_file(path):
    categories = {}
    current_name = None
    current_id = None
    with open(path, 'r') as f:
        for raw in f:
            line = raw.strip()
            if not line:
                continue
            if line.startswith('//'):
                text = line.strip('/').strip()
                if '====' in text or text.startswith('Index'):
                    continue
                if text:
                    current_name = text
                continue
            if re.fullmatch(r'\d+', line):
                current_id = int(line)
                categories[current_id] = {'name': current_name, 'items': []}
                continue
            if line.startswith('end'):
                current_id = None
                continue
            if current_id is not None:
                # split by quotes to extract name
                parts = line.split('"')
                name = parts[1] if len(parts) > 1 else 'Unknown'
                numbers = parts[0].strip().split()
                if len(numbers) >= 5:
                    index = int(numbers[0])
                    width = int(numbers[3])
                    height = int(numbers[4])
                    categories[current_id]['items'].append({
                        'index': index,
                        'name': name,
                        'width': width,
                        'height': height
                    })
    return categories


def parse_shop_file(path):
    entries = []
    with open(path, 'r') as f:
        for raw in f:
            line = raw.strip()
            if not line or line.startswith('//') or line.startswith('end'):
                continue
            base = line.split('//')[0]
            parts = base.replace(',', ' ').split()
            if len(parts) < 2:
                continue
            try:
                cat_id = int(parts[0])
                idx = int(parts[1])
            except ValueError:
                continue
            slotx = None
            sloty = None
            if len(parts) >= 9:
                if parts[7] != '*':
                    slotx = int(parts[7])
                if parts[8] != '*':
                    sloty = int(parts[8])
            entries.append({'category': cat_id, 'index': idx, 'x': slotx, 'y': sloty})
    return entries


def load_image_for(item):
    if Image is None:
        return None
    name = item['name'].replace(' ', '_')
    for ext in ('png', 'jpg', 'gif'):
        path = os.path.join('Data', 'Img', f"{name}.{ext}")
        if os.path.exists(path):
            img = Image.open(path)
            img = img.resize((item['width'] * CELL_SIZE, item['height'] * CELL_SIZE))
            return ImageTk.PhotoImage(img)
    return None


class ShopEditor(tk.Tk):
    def __init__(self):
        super().__init__()
        self.title('Shop Editor')
        self.items_data = parse_item_file(os.path.join('Data', 'item.txt'))
        self.category_var = tk.StringVar()
        self.category_combo = ttk.Combobox(self, textvariable=self.category_var, state='readonly')
        values = [f"{cid} - {v['name']}" for cid, v in sorted(self.items_data.items())]
        self.category_combo['values'] = values
        if self.category_combo['values']:
            self.category_combo.current(0)
        self.category_combo.pack(fill='x')
        self.category_combo.bind('<<ComboboxSelected>>', self.update_item_list)

        self.item_list = tk.Listbox(self, height=8)
        self.item_list.pack(fill='both')

        btn_frame = tk.Frame(self)
        btn_frame.pack(fill='x')
        tk.Button(btn_frame, text='Load Shop', command=self.load_shop).pack(side='left')
        tk.Button(btn_frame, text='Save Shop', command=self.save_shop).pack(side='left')
        tk.Button(btn_frame, text='Clear', command=self.clear_grid).pack(side='left')

        self.canvas = tk.Canvas(self, width=GRID_W*CELL_SIZE, height=GRID_H*CELL_SIZE, bg='white')
        self.canvas.pack(side='right', padx=5, pady=5)
        for x in range(GRID_W+1):
            self.canvas.create_line(x*CELL_SIZE, 0, x*CELL_SIZE, GRID_H*CELL_SIZE, fill='gray')
        for y in range(GRID_H+1):
            self.canvas.create_line(0, y*CELL_SIZE, GRID_W*CELL_SIZE, y*CELL_SIZE, fill='gray')
        self.canvas.bind('<Button-1>', self.place_item_event)
        self.canvas.bind('<Button-3>', self.remove_item_event)

        self.images = []
        self.grid_data = [[None for _ in range(GRID_W)] for _ in range(GRID_H)]
        self.update_item_list()

    def get_selected_category_id(self):
        sel = self.category_combo.get()
        return int(sel.split(' ')[0])

    def update_item_list(self, event=None):
        cid = self.get_selected_category_id()
        self.item_list.delete(0, tk.END)
        for it in self.items_data[cid]['items']:
            self.item_list.insert(tk.END, f"{it['index']:03d} {it['name']}")

    def place_item_event(self, event):
        selection = self.item_list.curselection()
        if not selection:
            return
        cid = self.get_selected_category_id()
        item = self.items_data[cid]['items'][selection[0]]
        gx = event.x // CELL_SIZE
        gy = event.y // CELL_SIZE
        if gx + item['width'] > GRID_W or gy + item['height'] > GRID_H:
            messagebox.showerror('Error', 'Item outside grid')
            return
        for y in range(gy, gy + item['height']):
            for x in range(gx, gx + item['width']):
                if self.grid_data[y][x] is not None:
                    messagebox.showerror('Error', 'Space occupied')
                    return
        img = load_image_for(item)
        if img:
            obj = self.canvas.create_image(gx*CELL_SIZE, gy*CELL_SIZE, anchor='nw', image=img)
            self.images.append(img)
        else:
            obj = self.canvas.create_rectangle(gx*CELL_SIZE, gy*CELL_SIZE,
                                               (gx+item['width'])*CELL_SIZE,
                                               (gy+item['height'])*CELL_SIZE,
                                               fill='skyblue')
            self.canvas.create_text((gx+item['width']/2)*CELL_SIZE,
                                    (gy+item['height']/2)*CELL_SIZE,
                                    text=item['name'][:4], font=('Arial', 8))
        entry = {'category': cid, 'index': item['index'],
                 'x': gx, 'y': gy, 'width': item['width'], 'height': item['height'],
                 'obj': obj}
        for y in range(gy, gy + item['height']):
            for x in range(gx, gx + item['width']):
                self.grid_data[y][x] = entry

    def remove_item_event(self, event):
        gx = event.x // CELL_SIZE
        gy = event.y // CELL_SIZE
        entry = self.grid_data[gy][gx]
        if not entry:
            return
        self.canvas.delete(entry['obj'])
        for y in range(entry['y'], entry['y'] + entry['height']):
            for x in range(entry['x'], entry['x'] + entry['width']):
                self.grid_data[y][x] = None

    def clear_grid(self):
        self.canvas.delete('all')
        for x in range(GRID_W+1):
            self.canvas.create_line(x*CELL_SIZE, 0, x*CELL_SIZE, GRID_H*CELL_SIZE, fill='gray')
        for y in range(GRID_H+1):
            self.canvas.create_line(0, y*CELL_SIZE, GRID_W*CELL_SIZE, y*CELL_SIZE, fill='gray')
        self.grid_data = [[None for _ in range(GRID_W)] for _ in range(GRID_H)]
        self.images.clear()

    def load_shop(self):
        path = filedialog.askopenfilename(title='Open shop', filetypes=[('Text','*.txt')])
        if not path:
            return
        self.clear_grid()
        for e in parse_shop_file(path):
            if e['x'] is None or e['y'] is None:
                continue
            cat = self.items_data.get(e['category'])
            if not cat:
                continue
            item = next((i for i in cat['items'] if i['index']==e['index']), None)
            if not item:
                continue
            # emulate placing
            gx, gy = e['x'], e['y']
            if gx + item['width'] > GRID_W or gy + item['height'] > GRID_H:
                continue
            occupied = False
            for y in range(gy, gy + item['height']):
                for x in range(gx, gx + item['width']):
                    if self.grid_data[y][x] is not None:
                        occupied = True
                        break
                if occupied:
                    break
            if occupied:
                continue
            img = load_image_for(item)
            if img:
                obj = self.canvas.create_image(gx*CELL_SIZE, gy*CELL_SIZE, anchor='nw', image=img)
                self.images.append(img)
            else:
                obj = self.canvas.create_rectangle(gx*CELL_SIZE, gy*CELL_SIZE,
                                                   (gx+item['width'])*CELL_SIZE,
                                                   (gy+item['height'])*CELL_SIZE,
                                                   fill='skyblue')
                self.canvas.create_text((gx+item['width']/2)*CELL_SIZE,
                                        (gy+item['height']/2)*CELL_SIZE,
                                        text=item['name'][:4], font=('Arial', 8))
            entry = {'category': e['category'], 'index': item['index'],
                     'x': gx, 'y': gy, 'width': item['width'], 'height': item['height'],
                     'obj': obj}
            for y in range(gy, gy + item['height']):
                for x in range(gx, gx + item['width']):
                    self.grid_data[y][x] = entry
        messagebox.showinfo('Loaded', f'Shop loaded from {path}')

    def save_shop(self):
        path = filedialog.asksaveasfilename(defaultextension='.txt', title='Save shop')
        if not path:
            return
        with open(path, 'w') as f:
            f.write('// Shop generated by ShopEditor\n')
            f.write('//Index\tLevel\tDur\tSkill\tLuck\tOption\tExcOp\tSlotX\tSlotY\tComment\n')
            used = set()
            for y in range(GRID_H):
                for x in range(GRID_W):
                    entry = self.grid_data[y][x]
                    if entry and entry not in used:
                        used.add(entry)
                        name = next((i['name'] for i in self.items_data[entry['category']]['items'] if i['index']==entry['index']), 'Item')
                        f.write(f"{entry['category']:02d},{entry['index']:03d}\t0\t*\t0\t0\t0\t0\t{entry['x']}\t{entry['y']}\t//{name}\n")
            f.write('end\n')
        messagebox.showinfo('Saved', f'Shop saved to {path}')


if __name__ == '__main__':
    app = ShopEditor()
    app.mainloop()

File: shop_editor.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using System.Windows.Forms;

public class ShopEditor : Form
{
    private ComboBox categoryCombo;
    private ListBox itemList;
    private DataGridView grid;
    private Button loadButton;
    private Button saveButton;
    private Dictionary<int, Category> categories;
    private Item selectedItem;
    private GridEntry[,] gridData = new GridEntry[15,8];

    public ShopEditor()
    {
        Text = "Shop Editor";
        Width = 800;
        Height = 600;
        InitComponents();
        categories = ItemParser.Parse("Data/item.txt");
        foreach(var kv in categories.OrderBy(k=>k.Key))
        {
            categoryCombo.Items.Add($"{kv.Key} - {kv.Value.Name}");
        }
        if(categoryCombo.Items.Count>0) categoryCombo.SelectedIndex = 0;
        UpdateItemList();
    }

    private void InitComponents()
    {
        categoryCombo = new ComboBox { Dock = DockStyle.Top, DropDownStyle = ComboBoxStyle.DropDownList };
        categoryCombo.SelectedIndexChanged += (s,e)=> UpdateItemList();
        itemList = new ListBox { Dock = DockStyle.Top, Height = 120 };
        itemList.SelectedIndexChanged += ItemList_SelectedIndexChanged;
        grid = new DataGridView {
            Dock = DockStyle.Fill,
            ColumnCount = 8,
            AllowUserToAddRows = false,
            AllowUserToDeleteRows = false,
            AllowUserToResizeColumns = false,
            AllowUserToResizeRows = false,
            MultiSelect = false,
            RowHeadersVisible = false,
            ColumnHeadersVisible = false,
            ReadOnly = true,
            SelectionMode = DataGridViewSelectionMode.CellSelect,
            BackgroundColor = Color.DimGray,
            EnableHeadersVisualStyles = false
        };
        grid.DefaultCellStyle.BackColor = Color.DimGray;
        grid.RowTemplate.Height = 32;
        grid.RowCount = 15;
        grid.CellClick += Grid_CellClick;
        grid.CellMouseDown += Grid_CellMouseDown;
        for(int i=0;i<8;i++)
            grid.Columns[i].Width = 32;
        loadButton = new Button { Text = "Load Shop" };
        loadButton.Click += (s,e)=>LoadShop();
        saveButton = new Button { Text = "Save Shop" };
        saveButton.Click += (s,e)=>SaveShop();
        var buttonPanel = new FlowLayoutPanel { Dock = DockStyle.Top, AutoSize = true };
        buttonPanel.Controls.Add(loadButton);
        buttonPanel.Controls.Add(saveButton);
        Controls.Add(grid);
        Controls.Add(buttonPanel);
        Controls.Add(itemList);
        Controls.Add(categoryCombo);
    }

    private void UpdateItemList()
    {
        itemList.Items.Clear();
        if(categoryCombo.SelectedIndex<0) return;
        int cid = GetSelectedCategoryId();
        foreach(var item in categories[cid].Items)
        {
            itemList.Items.Add($"{item.Index:D3} {item.Name}");
        }
    }

    private int GetSelectedCategoryId()
    {
        var text = categoryCombo.SelectedItem.ToString();
        var parts = text.Split(' ');
        int cid=0; int.TryParse(parts[0], out cid); return cid;
    }

    private void Grid_CellClick(object sender, DataGridViewCellEventArgs e)
    {
        if(e.RowIndex<0||e.ColumnIndex<0) return;
        if(selectedItem==null)
        {
            var entry = gridData[e.RowIndex,e.ColumnIndex];
            if(entry!=null) RemoveEntry(entry);
            return;
        }
        PlaceItem(e.ColumnIndex,e.RowIndex,selectedItem);
    }

    private void Grid_CellMouseDown(object sender, DataGridViewCellMouseEventArgs e)
    {
        if(e.Button==MouseButtons.Right && e.RowIndex>=0 && e.ColumnIndex>=0)
        {
            var entry = gridData[e.RowIndex,e.ColumnIndex];
            if(entry!=null) RemoveEntry(entry);
        }
    }

    private void PlaceItem(int gx, int gy, Item item)
    {
        if(gx+item.Width>8 || gy+item.Height>15) return;
        for(int y=gy;y<gy+item.Height;y++)
            for(int x=gx;x<gx+item.Width;x++)
                if(gridData[y,x]!=null) return;
        var entry = new GridEntry{Item=item,X=gx,Y=gy};
        for(int y=gy;y<gy+item.Height;y++)
            for(int x=gx;x<gx+item.Width;x++)
            {
                gridData[y,x]=entry;
                grid.Rows[y].Cells[x].Style.BackColor = Color.LightBlue;
                if(y==gy && x==gx)
                    grid.Rows[y].Cells[x].Value = item.Name.Substring(0,Math.Min(4,item.Name.Length));
                else
                    grid.Rows[y].Cells[x].Value = "";
            }
    }

    private void RemoveEntry(GridEntry entry)
    {
        for(int y=entry.Y;y<entry.Y+entry.Item.Height;y++)
            for(int x=entry.X;x<entry.X+entry.Item.Width;x++)
            {
                gridData[y,x]=null;
                grid.Rows[y].Cells[x].Value = null;
                grid.Rows[y].Cells[x].Style.BackColor = Color.DimGray;
            }
    }

    private void ItemList_SelectedIndexChanged(object sender, EventArgs ev)
    {
        var idx = itemList.SelectedIndex;
        if(idx>=0)
        {
            int cid = GetSelectedCategoryId();
            selectedItem = categories[cid].Items[idx];
        }
        else
        {
            selectedItem = null;
        }
    }

    private void LoadShop()
    {
        var dlg = new OpenFileDialog{Filter="Text|*.txt"};
        if(dlg.ShowDialog()!=DialogResult.OK) return;
        ClearGrid();
        foreach(var entry in ShopParser.Parse(dlg.FileName))
        {
            if(categories.TryGetValue(entry.Category,out var cat))
            {
                var item = cat.Items.FirstOrDefault(i=>i.Index==entry.Index);
                if(item!=null) PlaceItem(entry.X,entry.Y,item);
            }
        }
    }

    private void SaveShop()
    {
        var dlg = new SaveFileDialog{Filter="Text|*.txt"};
        if(dlg.ShowDialog()!=DialogResult.OK) return;
        using(var sw = new StreamWriter(dlg.FileName))
        {
            sw.WriteLine("// Shop generated by ShopEditor");
            sw.WriteLine("//Index\tLevel\tDur\tSkill\tLuck\tOption\tExcOp\tSlotX\tSlotY\tComment");
            HashSet<GridEntry> written = new HashSet<GridEntry>();
            for(int y=0;y<15;y++)
            for(int x=0;x<8;x++)
            {
                var entry = gridData[y,x];
                if(entry!=null && !written.Contains(entry))
                {
                    written.Add(entry);
                    sw.WriteLine($"{entry.Item.Category:D2},{entry.Item.Index:D3}\t0\t*\t0\t0\t0\t0\t{entry.X}\t{entry.Y}\t//{entry.Item.Name}");
                }
            }
            sw.WriteLine("end");
        }
    }

    private void ClearGrid()
    {
        for(int y=0;y<15;y++)
            for(int x=0;x<8;x++)
            {
                gridData[y,x]=null;
                grid.Rows[y].Cells[x].Value=null;
                grid.Rows[y].Cells[x].Style.BackColor = Color.DimGray;
            }
    }

    [STAThread]
    public static void Main()
    {
        Application.EnableVisualStyles();
        Application.Run(new ShopEditor());
    }
}

public class GridEntry
{
    public Item Item;
    public int X;
    public int Y;
}

public class Item
{
    public int Index;
    public int Width;
    public int Height;
    public string Name;
    public int Category;
}

public class Category
{
    public int Id;
    public string Name;
    public List<Item> Items = new List<Item>();
}

public static class ItemParser
{
    public static Dictionary<int,Category> Parse(string path)
    {
        var lines = File.ReadAllLines(path);
        var cats = new Dictionary<int,Category>();
        string catName = null;
        int catId = -1;
        foreach(var raw in lines)
        {
            var line = raw.Trim();
            if(line.StartsWith("//") && line.Contains("===="))
                continue;
            if(line.StartsWith("//"))
            {
                catName = line.Trim('/',' ','\t');
                continue;
            }
            if(line=="end")
            {
                catName=null; catId=-1; continue;
            }
            if(string.IsNullOrWhiteSpace(line)) continue;
            if(catId==-1 && int.TryParse(line,out catId))
            {
                var cat = new Category{Id=catId, Name=catName};
                cats[catId]=cat; continue;
            }
            if(catId!=-1)
            {
                var parts = line.Split('"');
                if(parts.Length<2) continue;
                string before = parts[0];
                string name = parts[1];
                var nums = before.Split(new char[]{' ','\t'},StringSplitOptions.RemoveEmptyEntries);
                if(nums.Length>=5)
                {
                    int idx=int.Parse(nums[0]);
                    int width=int.Parse(nums[3]);
                    int height=int.Parse(nums[4]);
                    cats[catId].Items.Add(new Item{Index=idx,Width=width,Height=height,Name=name,Category=catId});
                }
            }
        }
        return cats;
    }
}

public class ShopEntry
{
    public int Category;
    public int Index;
    public int X;
    public int Y;
}

public static class ShopParser
{
    public static IEnumerable<ShopEntry> Parse(string path)
    {
        foreach(var raw in File.ReadAllLines(path))
        {
            var line = raw.Trim();
            if(string.IsNullOrWhiteSpace(line) || line.StartsWith("//") || line=="end") continue;
            var basePart = line.Split(new[]{'/'},2)[0];
            var parts = basePart.Replace(',', ' ').Split(new char[]{' ','\t'},StringSplitOptions.RemoveEmptyEntries);
            if(parts.Length<9) continue;
            if(!int.TryParse(parts[0],out int cat)) continue;
            if(!int.TryParse(parts[1],out int idx)) continue;
            if(!int.TryParse(parts[7],out int x)) continue;
            if(!int.TryParse(parts[8],out int y)) continue;
            yield return new ShopEntry{Category=cat,Index=idx,X=x,Y=y};
        }
    }
}
